File: cpp/interp.cpp
#include "interp.hpp"

template<integer shape1, integer shape2>
void Interpolator::central(const arrType<scalar, shape1, shape2>& phi, scalar* phiF, integer index) {
    const Mesh& mesh = *this->mesh;
    integer p = mesh.owner(index);
    integer n = mesh.neighbour(index);
    scalar w = mesh.weights(index);
    //integer w = mesh.weights(index);
    for (integer i = 0; i < shape1; i++) {
        for (integer j = 0; j < shape2; j++) {
            phiF[i*phi.shape[2]+j] = phi(n, i, j)*(1-w) + phi(p, i, j)*w;
        }
    }
}

template<integer shape1, integer shape2>
void Interpolator::average(const arrType<scalar, shape1, shape2>& phi, scalar* phiF, integer index) {
    const Mesh& mesh = *this->mesh;
    integer p = mesh.owner(index);
    integer n = mesh.neighbour(index);
    scalar w = 0.5;
    //integer w = mesh.weights(index);
    for (integer i = 0; i < shape1; i++) {
        for (integer j = 0; j < shape2; j++) {
            phiF[i*phi.shape[2]+j] = phi(n, i, j)*(1-w) + phi(p, i, j)*w;
        }
    }
}

template<integer shape1>
void Interpolator::firstOrder(const arrType<scalar, shape1>& phi, const arrType<scalar, shape1, 3>& gradPhi, scalar *phiF, integer index, integer swap) {
    const Mesh& mesh = *this->mesh;
    integer p;
    if (swap) {
        p = mesh.neighbour(index);
    } else {
        p = mesh.owner(index);
    }
    for (integer i = 0; i < shape1; i++) {
        phiF[i] = phi(p, i);
    }
}

template<integer shape1>
void Interpolator::secondOrder(const arrType<scalar, shape1>& phi, const arrType<scalar, shape1, 3>& gradPhi, scalar *phiF, integer index, integer swap) {
    const Mesh& mesh = *this->mesh;
    integer p, n;
    if (swap) {
        n = mesh.owner(index);
        p = mesh.neighbour(index);
    } else {
        p = mesh.owner(index);
        n = mesh.neighbour(index);
    }
    for (integer i = 0; i < shape1; i++) {
        scalar phiC = phi(p, i);
        scalar phiD = phi(n, i);
        phiF[i] = phiC + (phiD-phiC)*mesh.linearWeights(swap, index);
        for (integer j = 0; j < 3; j++) {
            phiF[i] += mesh.quadraticWeights(swap, index, j)*gradPhi(p, i, j);
        }
    }
}

File: adFVM/interp.py
from . import config, parallel
from .field import Field#, faceExchange

import itertools
import numpy as np
from scipy import sparse as sparse

logger = config.Logger(__name__)

# central is second order with no grad
# dual should do second order with grad
# characteristic is extrapolated

# dual defined for np and ad
def central(phi, mesh):
    f = mesh.weights
    phiF = phi.extract(mesh.owner)*f + phi.extract(mesh.neighbour)*(-f+1)
    return phiF

def secondOrder(phi, gradPhi, mesh, swap):
    p, n = mesh.owner, mesh.neighbour
    if swap:
        n, p = p, n
    phiC, phiD = phi.extract(p), phi.extract(n)
    phiF = phiC + (phiD-phiC)*mesh.linearWeights[swap] 
    for i in range(0, phiC.shape[0]):
        phiF[i] += mesh.quadraticWeights[swap].dot(gradPhi.extract(p)[i])
    return phiF

def centralOld(phi, mesh):
    logger.info('interpolating {0}'.format(phi.name))
    factor = mesh.weights
    # for tensor
    if len(phi.dimensions) == 2:
        factor = np.reshape(factor, (-1, 1, 1))
    faceField = Field('{0}F'.format(phi.name), phi.field[mesh.owner]*factor + phi.field[mesh.neighbour]*(1.-factor), phi.dimensions)
    #faceField = Field('{0}F'.format(phi.name), phi.field[mesh.owner], phi.dimensions)
    # retain pattern broadcasting
    #if hasattr(mesh, 'origMesh'):
    #    faceField.field = ad.patternbroadcast(faceField.field, phi.field.broadcastable)
    return faceField

